output/cloud/expv2/collect.go

```diff
@@ -0,0 +1,156 @@
+	q.m.Unlock()
+}
+
+type collector struct {
+	bq    bucketQ
+	nowfn func() time.Time
+
+	aggregationPeriod time.Duration
+	waitPeriod        time.Duration
+
+	// we should no longer have to handle metrics that have times long in the past. So instead of a
+	// map, we can probably use a simple slice (or even an array!) as a ring buffer to store the
+	// aggregation buckets. This should save us a some time, since it would make the lookups and WaitPeriod
+	// checks basically O(1). And even if for some reason there are occasional metrics with past times that
+	// don't fit in the chosen ring buffer size, we could just send them along to the buffer unaggregated
+	timeBuckets map[int64]map[metrics.TimeSeries]metrics.Sink
+}
+
+func newCollector(aggrPeriod, waitPeriod time.Duration) (*collector, error) {
+	if aggrPeriod == 0 {
+		return nil, errors.New("aggregation period is not allowed to be zero")
+	}
+	return &collector{
+		bq:                bucketQ{},
+		nowfn:             time.Now,
+		timeBuckets:       make(map[int64]map[metrics.TimeSeries]metrics.Sink),
+		aggregationPeriod: aggrPeriod,
+		waitPeriod:        waitPeriod,
+	}, nil
+}
+
+// CollectSamples drain the buffer and collect all the samples.
+func (c *collector) CollectSamples(containers []metrics.SampleContainer) {
+	// Distribute all newly buffered samples into related buckets
+	for _, sampleContainer := range containers {
+		samples := sampleContainer.GetSamples()
+
+		for i := 0; i < len(samples); i++ {
+			c.collectSample(samples[i])
+		}
+	}
+
+	expired := c.expiredBuckets()
+	if len(expired) < 1 {
+		return
+	}
+	c.bq.Push(expired)
+}
+
+// DropExpiringDelay drops the waiting time for buckets
+// for the expiring checks.
+func (c *collector) DropExpiringDelay() {
+	c.waitPeriod = 0
+}
+
+func (c *collector) collectSample(s metrics.Sample) {
+	bucketID := c.bucketID(s.Time)
+
+	// Get or create a time bucket
+	bucket, ok := c.timeBuckets[bucketID]
+	if !ok {
+		bucket = make(map[metrics.TimeSeries]metrics.Sink)
+		c.timeBuckets[bucketID] = bucket
+	}
+
+	// Get or create the bucket's sinks map per time series
+	sink, ok := bucket[s.TimeSeries]
+	if !ok {
+		sink = newSink(s.Metric.Type)
+		bucket[s.TimeSeries] = sink
+	}
+
+	// TODO: we may consider to just pass
+	// the single value instead of the entire
+	// sample and save some memory
+	sink.Add(s)
+}
+
+func (c *collector) expiredBuckets() []timeBucket {
+	// Still too recent buckets
+	// where we prefer to wait a bit more
+	// then, hopefully, we can aggregate more samples before flushing.
+	bucketCutoffID := c.bucketCutoffID()
+
+	// Here, it avoids pre-allocation
+	// because it expects to be zero for most of the time
+	var expired []timeBucket //nolint:prealloc
+
+	// Mark as expired all aggregation buckets older than bucketCutoffID
+	for bucketID, seriesSinks := range c.timeBuckets {
+		if bucketID > bucketCutoffID {
+			continue
+		}
+
+		expired = append(expired, timeBucket{
+			Time:  c.timeFromBucketID(bucketID),
```

Comment:
I would probably just leave it as the `bucketID`. I am not certain but it seems to me that this is what we end up writing out later on?

Reply:
I will do this at the end of the chain, I prefer to have the support of the entire test suite so I'm sure I'm doing everything correctly, especially on the Protobuf side. Is it fine for you?